src/aliasthis.c

```diff
@@ -21,18 +21,376 @@
 #include "mtype.h"
 #include "declaration.h"
 #include "tokens.h"
+#include "enum.h"
+#include "template.h"
+#include "arraytypes.h"
+#include "expression.h"
 
-Expression *resolveAliasThis(Scope *sc, Expression *e)
+typedef Array<const char *> Strings;
+
+/**
+ * iterateAliasThis resolves alias this subtypes for `e` and applies it to `dg`.
+ * dg should return true, if appropriate subtype has been found. Otherwise it should returns 0.
+ * Also dg can get context through secong parameter `ctx`, which was passed through fourth argument of iterateAliasThis.
+ * `dg` can return result expression through `outexpr` parameter, and if it is not NULL, it will be pushed to `ret` array.
+ * At the first stage iterateAliasThis iterates direct "alias this"-es and pushes non-null `outexpr`-es to `ret`.
+ * If `dg` for one of direct "alias this"-es returns true then `iterateAliasThis` breaks at this stage and returns true
+ * through return value and returned by `dg` expression array through `ret`.
```

Comment:
`ret` is a pointer to the `Expression*` array.
User passes it to the `iterateAliasThis` call. `iterateAliasThis` passes to `dg` a ponter to `Expression`. `dg` writes result (if it is found) to this argument, and if `iterateAliasThis` takes non-null expression from `dg`, it pushes it to the `ret` array.
General idea of this function: Only it manages of subtype tree iterating:
now it walks over all subtype tree, however if we will want to change walkup algorithm, we will need only to change this function (not each alias this using).
Current alias this resolve algorithm is not final. 
At least one improvement I can offer now.
We may allow alias this declaration to hide same type alias this declarations of basetypes:

```
struct A
{
    int i;
    alias i this;
}

struct B
{
    int i;
    alias i this;
}

struct C
{
    A a;
    B b;
    int i;
    alias i this;
    alias a this;
    alias b this;
}

void test()
{
    int a = C(); //now this code raises an error: C.i vs C.a.i vs C.b.i
                      //however we may allow to hide C.i of basetype "alias intSymbol this" declarations.
}
```

To make change like this we need only to change `iterateAliasThis` definition, and do not change >9000 alias this related fragments, because each those fragments (for example overloaded operator resolving) doesn't know about subtype tree walkup order. This fragment only defines corresponding delegate which know how to proccess one alias this canditate.